### PR TITLE
Perfetto connector

### PR DESCRIPTION
**Pull request by jrmadsen**
- alternative implementation to https://github.com/kokkos/kokkos-tools/pull/124
  - Has some configuration options
  - Supports both in-process and system backends

### CODE REVIEW

File: profiling/perfetto-connector/libperfetto-connector.cpp [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+      fprintf(stderr, "Error opening %s!\n", _fname.c_str());
+  }
+
+  printf("Done\n");
+  printf("-----------------------------------------------------------\n");
+  fflush(stdout);
+}
+
+void kokkosp_begin_parallel_for(const char* name, const uint32_t, uint64_t*) {
+  TRACE_EVENT_BEGIN("kokkos.kernels", perfetto::StaticString(name));
```
Comment (dalg24) [inline profiling/perfetto-connector/libperfetto-connector.cpp]: Not distinguishing `parallel_for` from `parallel_reduce` and `parallel_scan`?
I don't necessary have an issue with it.  Just wondering if that is by design.

File: profiling/perfetto-connector/libperfetto-connector.cpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+  tracing_session = perfetto::Tracing::NewTrace();
+  tracing_session->Setup(cfg);
+  tracing_session->StartBlocking();
+}
+
+void kokkosp_finalize_library() {
+  printf("-----------------------------------------------------------\n");
+  printf("KokkosP: Finalizing Perfetto Connector... ");
+  fflush(stdout);
```
Comment (dalg24) [inline profiling/perfetto-connector/libperfetto-connector.cpp]: Aww man I was so pleased with the environment parsing and now we are back to the good ole printf?

File: profiling/perfetto-connector/libperfetto-connector.cpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    perfetto::Category("kokkos.regions")
+        .SetDescription("Kokkos Region Events"));
+PERFETTO_TRACK_EVENT_STATIC_STORAGE();
+
+namespace {
+auto tracing_session = std::unique_ptr<perfetto::TracingSession>{};
+
+template <typename Tp>
+auto get_env(std::string_view _env_name, Tp _default) {
+  const char* _env_val = std::getenv(_env_name.data());
```
Comment (dalg24) [inline profiling/perfetto-connector/libperfetto-connector.cpp]: I think that is potentially a bug.  `std::getenv` expects null-terminated character string.
I would pass a `std::string` and use `c_str()`.  This is not performance critical.

**Review by dalg24 — COMMENTED** [review]
Looks mostly good besides the `string_view` thing

File: profiling/perfetto-connector/libperfetto-connector.cpp (hunk 3 above)
Comment (jrmadsen) [reply profiling/perfetto-connector/libperfetto-connector.cpp]: Good point

File: profiling/perfetto-connector/libperfetto-connector.cpp (hunk 1 above)
Comment (jrmadsen) [reply profiling/perfetto-connector/libperfetto-connector.cpp]: Nope not by design, just inherited this from David's. It's a fair point though and easy to fix

File: profiling/perfetto-connector/libperfetto-connector.cpp (hunk 2 above)
Comment (jrmadsen) [reply profiling/perfetto-connector/libperfetto-connector.cpp]: 🤣 I just tweaked this from David's. I used to never use it but, over time, I've come to appreciate it more. I'd replace it with std::format in a heartbeat.

File: profiling/perfetto-connector/Makefile [hunk 4]
```diff
@@ -0,0 +1,18 @@
+
+CXX := g++
+CXXFLAGS := -O3 -std=c++17 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -W -Wall
+
+default: all
+
+all: library strip
+
+library: libperfetto-connector.cpp
```
Comment (dalg24) [inline profiling/perfetto-connector/Makefile]: It is rebuilding each time
```suggestion
all: libperfetto-connector.so strip

libperfetto-connector.so: libperfetto-connector.cpp
```

File: profiling/perfetto-connector/Makefile [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+	$(CXX) $(CXXFLAGS) -I. libperfetto-connector.cpp perfetto.cc -shared -o libperfetto-connector.so
+
+strip: libperfetto-connector.so
+	strip -w -K kokkosp_* libperfetto-connector.so
```
Comment (dalg24) [inline profiling/perfetto-connector/Makefile]: Doesn't work on OSX
```
strip -w -K kokkosp_* libperfetto-connector.so
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip: unrecognized option: -w
Usage: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip [-AnuSXx] [-] [-d filename] [-s filename] [-R filename] [-o output] file [...]
make: *** [strip] Error 1
```

File: profiling/perfetto-connector/libperfetto-connector.cpp [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+  perfetto::TrackEvent::Register();
+
+  tracing_session = perfetto::Tracing::NewTrace();
+  tracing_session->Setup(cfg);
+  tracing_session->StartBlocking();
+}
+
+void kokkosp_finalize_library() {
+  std::cout << "-----------------------------------------------------------\n"
+            << "KokkosP: Finalizing Perfetto Connector... " << std::flush;
```
Comment (dalg24) [inline profiling/perfetto-connector/libperfetto-connector.cpp]:
```suggestion
            << "KokkosP: Finalizing Perfetto Connector... " << std::endl;
```

```
17: -----------------------------------------------------------
17: KokkosP: Finalizing Perfetto Connector... [214.151] perfetto.cc:47237       Scraping SMB for producer 1
17: [214.156] perfetto.cc:38681       Stopping data source 1
17: [214.156] perfetto.cc:38721       Ending async stop of data source 1
17: [214.156] perfetto.cc:47237       Scraping SMB for producer 1
17: Reading trace data... Writing 'kokkosp-perfetto.pftrace'... Done

**Review by dalg24 — COMMENTED** [review]
Looks good other than the couple comments

**Comment by dalg24** [comment]
Well looks like we still have the two branches so you need to target `develop`

File: profiling/perfetto-connector/Makefile (hunk 5 above)
Comment (jrmadsen) [reply profiling/perfetto-connector/Makefile]: Fixed

**Review by dalg24 — APPROVED** [review]
`strip` rule is being re-run every time but I can live with it.

I just need to check with @crtrott regarding the copyright for the Perfetto files before we merge.

**Comment by crtrott** [comment]
Hey this looks good, trying to figure out the license situation. We may wanna add something like this statement from LLVM:

```
==============================================================================
Software from third parties included in the LLVM Project:
==============================================================================
The LLVM Project contains third party software which is under different license
terms. All such code will be identified clearly using at least one of two
mechanisms:
1) It will be in a separate directory tree with its own `LICENSE.txt` or
   `LICENSE` file at the top containing the specific license and restrictions
   which apply to that software, or
2) It will contain specific license and restriction terms at the top of every
   file.
```

But we will deal with it.

**Comment by jrmadsen** [comment]
> `strip` rule is being re-run every time but I can live with it.

Not sure that can be fixed. Every time it runs, it touches libperfetto-connector.so so it thinks it's out-of-date on the next run.

**Comment by dalg24** [comment]
> > `strip` rule is being re-run every time but I can live with it.
> 
> Not sure that can be fixed. Every time it runs, it touches libperfetto-connector.so so it thinks it's out-of-date on the next run.

Could be part of the same rule that produce the library I suppose.  But no big deal.

**Comment by jrmadsen** [comment]
@dalg24 @crtrott I moved perfetto.h and perfetto.cc into a perfetto subdirectory, added the LICENSE from their repo, and also copied it into the top of the header.

**Comment by jrmadsen** [comment]
> Could be part of the same rule that produce the library I suppose. But no big deal.

Yeah I actually made it separate just in case, for some reason, someone wanted to build and not strip it. Although I guess they could just remove the line if the wanted

**Comment by jrmadsen** [comment]
Ping

**Comment by jrmadsen** [comment]
@crtrott @dalg24 is the license stuff resolved?

**Comment by crtrott** [comment]
We have the license stuff resolved !! 

I.e. we finally got license change to Apache2 with LLVM exceptions approved, including the wording about containing code with other licenses in subdirectory.

**Comment by crtrott** [comment]
We are currently doing the license update right now in core and kernels.

**Review by crtrott — APPROVED** [review]

**Comment by jrmadsen** [comment]
BTW, it should probably be noted in the chrome-tracing that this is effectively the newer version of that tool.

Do we want to apply formatting to the external perfetto files or exclude it from formatting?

**Pull request merged**